lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -0,0 +1,236 @@
+            ):
+                continue
+            c = plugin.disable()
+            for key in c.keys():
+                configs[key] += c[key]
+
+        self._update_keystore_and_reload(configs["to_remove_keystore"], remove_keys=True)
+        needs_restart = self._config.config_delete_list(configs["to_remove_opensearch"])
+        return self._config.config_put_list(configs["to_add_opensearch"]) or needs_restart
+
+    def status(self) -> Dict[PluginState, List[str]]:
+        """Returns a dict summarizing the status of each plugin.
+
+        It can be converted to str and used in the message status or for logging.
+        """
+        full_status = {}
+        for plugin_name, plugin in self.plugins.items():
+            stat = plugin.status
+            if isinstance(full_status[stat], list):
+                full_status[stat].append(plugin_name)
+            else:
+                full_status[stat] = [plugin_name]
+        return full_status
+
+    def plugins_need_upgrade(self) -> List[OpenSearchPlugin]:
+        """Returns a list of plugins that need upgrade."""
+        return [name for name, obj in self.plugins.items() if obj.needs_upgrade]
+
+    def _is_plugin_relation_set(self, relation_name: str) -> bool:
+        """Returns True if a relation is expected and it is set."""
+        if not relation_name:
+            return True
+        return len(self._charm.framework.model.relations[relation_name] or {}) > 0
+
+    def _update_keystore_and_reload(
+        self, keystore: Dict[str, str], remove_keys: bool = False
+    ) -> None:
+        """Updates the keystore value (adding or removing) and reload."""
+        if not keystore:
+            return
+        try:
+            for key, value in keystore.items():
+                if remove_keys:
+                    self._remove_from_keystore(key)
+                else:
+                    self._add_to_keystore(key, value)
+            # Now, reload the security settings and return if opensearch needs restart
+            post = self._distro.request("POST", "_nodes/reload_secure_settings")
+            logger.debug(f"_update_keystore_and_reload: response received {post}")
+        except OpenSearchKeystoreError as ek:
+            raise ek
+        except Exception as e:
+            logger.exception(e)
+            raise OpenSearchPluginError("Unknown error during keystore reload")
+        if post["status"] < 200 or post["status"] >= 300:
+            raise OpenSearchPluginError("Error while processing _nodes reload")
+
+    def _add_plugin(self, plugin: str) -> bool:
+        """Add a plugin to this node. Restart must be managed in separated."""
+        try:
+            args = f"install --batch {plugin}"
+            self._distro.run_bin("opensearch-plugin", args)
+        except OpenSearchCmdError as e:
+            if "already exists" in str(e):
+                return
+            raise OpenSearchPluginError(f"Failed to install plugin {plugin}: " + str(e))
```

Comment:
1. If we go with the route of having plugin manager running an `_install` per plugin: we could incorporate the code of these methods in the `_install` methods without having them separate

2. Create a couple of more exceptions: `OpenSearchPluginInstall/RemoveError(OpenSearchPluginError)` those are the ones you'll raise below


```suggestion
        """Add a plugin to this node. Restart must be managed in separated."""
        try:
            self._distro.run_bin("opensearch-plugin", f"install --batch {plugin}")
        except OpenSearchCmdError as e:
            if "already exists" in str(e):
                return
            raise OpenSearchPluginInstallError(e)
```

Reply:
install method has some pre-checks before actually calling the ```_add_plugin```